### PR TITLE
Updating sympy.stats.crv_types for extreme value distributions

### PR DESCRIPTION
**Pull request by czgdp1807**
#### References
[1] https://en.wikipedia.org/wiki/Generalized_extreme_value_distribution#Link_to_Fr%C3%A9chet,_Weibull_and_Gumbel_families
[2] https://en.wikipedia.org/wiki/Gumbel_distribution
[3] Conitnued https://github.com/sympy/sympy/pull/16381
[4] https://prism.ucalgary.ca/bitstream/handle/11023/796/ucalgary_2013_ji_chaoqun.pdf?sequence=4&isAllowed=y

#### Brief description of what is fixed or changed
PDF for GumbelDistribution has been corrected with changes made at appropriate places(docs, tests, code)
CDF for GumbelDistribution has been added. Associated test have also been added

#### Other comments
N/A

#### Future Work
The following tasks will be done for `1.5`, mentioned in the order of decreasing priority,

Adding `GumbelMinimumDistribution`, `GumbelMaximumDistribution` and modifying `Gumbel` function.

#### Release Notes

* stats
  * Gumbel distribution(minimum and maximum) have been added to `sympy.stats`

### CODE REVIEW

**Comment by czgdp1807** [comment]
@sidhantnagpal and @sylee957 Please see this. I have closed #16381 due to git issues caused due to rebasing. Apologies for inconvenience caused.

**Comment by sidhantnagpal** [comment]
The base branch should be 1.4.

**Comment by czgdp1807** [comment]
@sidhantnagpal Please read the future work section of the main comment of this PR. This is for `1.5`.

**Comment by sidhantnagpal** [comment]
It looks like the Future Work section was added afterwards. By the way, since you opened this prior to #16529, the description here would be obsolete once 1.4 is merged into master (ie after the release). I'd recommend updating the description and title to bring it more in line with the exact changes you intend to make.

**Comment by sylee957** [comment]
Technically, your commit hashes are different as you can see, probably because of your attempt to edit commit in one of the clones, in different way.

https://github.com/sympy/sympy/pull/16529/commits/f1b8d1d4373486681ca8bd9ceedf433470b7b2a6 https://github.com/sympy/sympy/pull/16528/commits/39dc1d66fc90bddb2db9cc41679b2e1f55f4bd0e

Actually, I think that the github would automatically have closed this PR, if the merged one had the identical commit as this one, and wouldn't have shown some duplicate diffs, surprisingly.

**Comment by czgdp1807** [comment]
@sidhantnagpal I have updated the description and the title.

**Comment by czgdp1807** [comment]
@sidhantnagpal and @sylee957 Please review the changes I have made for adding `GumbelDistributionMinimum` and `GumbelDistributionMaximum`. 
I also want to know what options to pass to `latex()` to correct the docstrings, if there are any problems with it.
If there any suggestions to apply DRY in this PR, then please let me know.

**Comment by czgdp1807** [comment]
@sidhantnagpal [This](https://travis-ci.org/sympy/sympy/jobs/515094913) job is failing. 
Please suggest some way to resolve it. I cannot understand if it's related to my changes.

File: sympy/stats/crv_types.py [hunk 1]
```diff
@@ -1437,40 +1437,80 @@ def GammaInverse(name, a, b):
...
+
+    with :math:`x \in [ - \infty, \infty ]`.
 
     Parameters
     ==========
 
     mu: Real number, 'mu' is a location
     beta: Real number, 'beta > 0' is a scale
+    for_min: Boolean, optional, False, by default
+             For enabling the minimum distribution
```
Comment (sylee957) [inline sympy/stats/crv_types.py]:
```suggestion
        For enabling the minimum distribution
```

**Review by sylee957 — COMMENTED** [review]
```
Warning, treated as error:
/home/travis/build/sympy/sympy/sympy/stats/crv_types.py:docstring of sympy.stats.Gumbel:9:Unexpected indentation.
```
It is a sphinx error while building your doc.
I see you have some inconsistent indentation, one for `Parameters`, and two others for `math::` directive,
but I'm only sure about fixing parameters, and not very sure about correctly formatting math in multiple lines, yet.

If my suggestion doesn't fix your issue, you may also have to take a look at your math directives, as well.
Generally, you would have to strictly make indentations, 4 spaces, consistently for each paragraph.

**Comment by czgdp1807** [comment]
@sylee957 Thanks for your suggestions. I have committed the same.
Any reviews from your side regarding the code that I have added for completing task 1 of future work mentioned in the description.

> 1. Adding `GumbelMinimumDistribution`, `GumbelMaximumDistribution` and modifying `Gumbel` function.

**Comment by czgdp1807** [comment]
@jksuom @sidhantnagpal Any comments on [this](https://github.com/sympy/sympy/pull/16528#issuecomment-478948661) from your side?

File: sympy/stats/crv_types.py [hunk 2]
```diff
@@ -1441,46 +1441,78 @@ def GammaInverse(name, a, b):
     return rv(name, GammaInverseDistribution, (a, b))
 
 #-------------------------------------------------------------------------------
-# Gumbel distribution --------------------------------------------------------
+# Gumbel distribution (Maximum and Minimum)--------------------------------------------------------
 
 
-class GumbelDistribution(SingleContinuousDistribution):
+class GumbelDistributionMaximum(SingleContinuousDistribution):
     _argnames = ('beta', 'mu')
```
Comment (sidhantnagpal) [inline sympy/stats/crv_types.py]: Would it be better to define an additional argument, instead of having separate classes?

Comment (czgdp1807) [reply sympy/stats/crv_types.py]: Yeah, we can do that. I defined separate classes for decoupling the properties of `GumbelDistributionMaximum` & `GumbelDistributionMinimum`. For example, if `integrate` is unable handle some property, say for instance mean or variance or skewness, for `GumbelDistributionMaximum` but is able handle the same property for `GumbelDistributionMinimum` then we can solve the problem separately. If we will use the same class then any issue for one case will affect both the cases simultaneously. 
Waiting for your comments.

Comment (czgdp1807) [reply sympy/stats/crv_types.py]: What do you think @sidhantnagpal ?

Comment (sidhantnagpal) [reply sympy/stats/crv_types.py]: I am not certain about changing the name of the class. CC @Upabjojr

Comment (Upabjojr) [reply sympy/stats/crv_types.py]: If the distributions are the same, we should keep only one class.

Consider using if-clauses as a workaround (if integration fails, try another way).

Anyway, If integration fails on the sign-symmetry of the function, maybe it should be reported as a bug of the integrator.

**Comment by czgdp1807** [comment]
Please check [the computation by wolfram alpha for charateristic function of Minimum Gumbel Distribution](https://www.wolframalpha.com/input/?i=integrate++(exp(-exp((-mu+%2B+x)%2Fbeta)+%2B+(-mu+%2B+x)%2Fbeta)%2Fbeta)*exp(I*t*x)*dx+from+x%3D-inf+to+inf).
I don't think there is any result for it. We should return unevaluated integrals as is done by wolfram alpha. What do you say?

**Comment by sylee957** [comment]
In Mathematica, it evaluates into gamma function
```mathematica
In[93]:= PDF[ExtremeValueDistribution[\[Mu],\[Beta]],x]
Out[93]= E^(-E^(((-x+\[Mu])/\[Beta]))+(-x+\[Mu])/\[Beta])/\[Beta]

In[94]:= PDF[GumbelDistribution[\[Mu],\[Beta]],x]
Out[94]= E^(-E^(((x-\[Mu])/\[Beta]))+(x-\[Mu])/\[Beta])/\[Beta]

In[95]:= CharacteristicFunction[ExtremeValueDistribution[\[Mu],\[Beta]],x]
Out[95]= E^(I x \[Mu]) Gamma[1-I x \[Beta]]

In[96]:= CharacteristicFunction[GumbelDistribution[\[Mu],\[Beta]],x]
Out[96]= E^(I x \[Mu]) Gamma[1+I x \[Beta]]
```
Also, you may try alternative query than calling integral in Wolfram Alpha, like
https://www.wolframalpha.com/input/?i=characteristic+function+of+gumbel+distribution

I'm not sure if the fourier-type integral can succeed in SymPy or Mathematica, but if the integral fails in SymPy, you may also add some TODO note about it.

**Comment by czgdp1807** [comment]
Thanks @sylee957

**Comment by sylee957** [comment]
Besides, I see Wolfram defines minimal and maximal gumbel distribution as

https://reference.wolfram.com/language/ref/GumbelDistribution.html
and
https://reference.wolfram.com/language/ref/ExtremeValueDistribution.html

I don't know the behind-the-scenes idea about how they had named like that, though.

**Comment by czgdp1807** [comment]
I am doubtful that they are using direct results. As the same system is unable to solve the integral. I don't know if I passed the incorrect formula to the integral. Can you verify that?

**Comment by sylee957** [comment]
image
I think that the integral succeeds with assumptions specified, but it took more than a minute to compute those integrals, while calling `CharacteristicFunction` only took a second

So, with similar assumptions, does it succeed in SymPy?

**Comment by czgdp1807** [comment]
**Code**
```python
from sympy.stats import Gumbel, density
from sympy import integrate, symbols, exp, I, S
mu, beta = symbols('mu, beta', positive=True)
t, x = symbols('t, x', real=True)
G = Gumbel('G', beta, mu)
f = density(G)(x)
z = integrate(f * exp(I*t*x), (x, S.Reals))
print(z)
```
**Output**
```
exp(mu/beta)*Integral(exp(-x/beta)*exp(-exp(mu/beta)*exp(-x/beta))*exp(I*t*x), (x, -oo, oo))/beta
```

**Comment by czgdp1807** [comment]
@Upabjojr @sidhantnagpal Please review the changes I have made, so that I can move on to task 2.

File: sympy/stats/tests/test_continuous_rv.py [hunk 3]
```diff
@@ -631,9 +631,12 @@ def test_gumbel():
     beta = Symbol("beta", positive=True)
     mu = Symbol("mu")
     x = Symbol("x")
+    y = Symbol("y")
     X = Gumbel("x", beta, mu)
+    Y = Gumbel("y", beta, mu, minimum=True)
     assert str(density(X)(x)) == 'exp(-exp(-(-mu + x)/beta) - (-mu + x)/beta)/beta'
-    assert cdf(X)(x) == exp(-exp((mu - x)/beta))
+    assert str(density(Y)(y)) == 'exp(-exp((-mu + y)/beta) + (-mu + y)/beta)/beta'
+    assert str(cdf(X)(x)) == 'exp(-exp(-(-mu + x)/beta))'
```
Comment (Upabjojr) [inline sympy/stats/tests/test_continuous_rv.py]: string comparison is not recommended

Comment (czgdp1807) [reply sympy/stats/tests/test_continuous_rv.py]: I have `.expand()` as nothing else was working out.

**Comment by czgdp1807** [comment]
So now, I will proceed with other GEV distributions to have the min-max choice.

**Comment by czgdp1807** [comment]
The work on this PR is complete. If possible please merge it.

File: sympy/stats/crv_types.py [hunk 4]
```diff
@@ -1563,58 +1563,78 @@ def GammaInverse(name, a, b):
...
+    with :math:`x \in [ - \infty, \infty ]`.
+
     Parameters
     ==========
 
-    mu: Real number, 'mu' is a location
-    beta: Real number, 'beta > 0' is a scale
+    mu : Real number, 'mu' is a location
+    beta : Real number, 'beta > 0' is a scale
+    minimum : Boolean, by default, False, for enabling minimum distribution
```
Comment (sidhantnagpal) [inline sympy/stats/crv_types.py]: The default corresponds to maximum distribution by your setting.

Comment (czgdp1807) [reply sympy/stats/crv_types.py]: Is it needed to be changed to minimum distribution by default? AFAIK, the original tests pass on keeping the maximum distribution as default. If minimum is set to True then they will fail. Am I wrong somewhere?

Comment (sidhantnagpal) [reply sympy/stats/crv_types.py]: >    minimum : Boolean, by default, False, for enabling minimum distribution

This looks a bit misleading. It reads as if the default is for enabling minimum distribution.

Comment (czgdp1807) [reply sympy/stats/crv_types.py]: I have made the change please take a look if now its clear. Resolve this conversation if you are in agreement of the change. Thanks

**Comment by Upabjojr** [comment]
I don't see any flaw in the code. @sidhantnagpal , you can decide when to merge.

**Comment by sidhantnagpal** [comment]
Looks good to me.

**Pull request merged**